timelock-agent/src/main/java/com/palantir/atlasdb/timelock/paxos/LeadershipCoordinatorFactory.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.timelock.paxos;
+
+import com.palantir.leader.LeaderElectionService;
+import com.palantir.leader.proxy.LeadershipCoordinator;
+import java.io.Closeable;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class LeadershipCoordinatorFactory implements Closeable {
+    private final Map<LeaderElectionService, LeadershipCoordinator> leadershipCoordinators = new ConcurrentHashMap<>();
+
+    public LeadershipCoordinator create(LeaderElectionService leaderElectionService) {
+        return leadershipCoordinators.computeIfAbsent(
+                leaderElectionService, _u -> LeadershipCoordinator.create(leaderElectionService));
+    }
+
+    @Override
+    public void close() {
+        leadershipCoordinators.keySet().forEach(k -> {
+            LeadershipCoordinator leadershipCoordinator = leadershipCoordinators.remove(k);
+            if (leadershipCoordinator != null) {
+                leadershipCoordinator.close();
+            }
+        });
+    }
```

Comment:
This is done to get the tests to work - the tests trigger TimelockAgent#shutdown which shuts the the tryToGainLeadership thread. If the leadershipTokenRef at this time is null, there is no mechanism for the proxy to schedule tryToGainLeadership again.

This behavior is the same as before, the only difference is that earlier an AwaitingLeadershipProxy instance was created per client. Since each test generates a random client and a new client creates a new instance of AwaitingLeadershipProxy, this would never crop up in the tests.